Modules/CPackDeb.cmake

```diff
@@ -674,18 +704,26 @@ if(CPACK_DEB_PACKAGE_COMPONENT)
   get_component_package_name(CPACK_DEBIAN_PACKAGE_NAME "${CPACK_DEB_PACKAGE_COMPONENT}")
 endif()
 
+if (NOT CPACK_DEBIAN_PACKAGE_GENERATE_SHLIBS_POLICY)
+  set(CPACK_DEBIAN_PACKAGE_GENERATE_SHLIBS_POLICY "=")
+endif()
+
 # Generate shlibs file
-foreach(_FILE ${CPACK_DEB_SHARED_OBJECT_FILES})
-  get_filename_component(_LIBNAME ${_FILE} NAME_WE)
-  get_filename_component(_LIBEXT ${_FILE} EXT)
+if(NOT DEFINED CPACK_DEBIAN_PACKAGE_GENERATE_SHLIBS)
+  set(CPACK_DEBIAN_PACKAGE_GENERATE_SHLIBS 1)
+endif()
 
-  # For now we just assume that anything after '+' or '-' in a library version is
-  # a build metadata which we do not want to be presented in dependency version
-  string(REGEX MATCH "so\\.([0-9]+\\.[0-9]+)([^+-]*)" _TARGET_VERSION ${_LIBEXT})
-  list(APPEND CPACK_DEBIAN_PACKAGE_SHLIBS_LIST "${_LIBNAME} ${CMAKE_MATCH_1} ${CPACK_DEBIAN_PACKAGE_NAME} (>= ${CMAKE_MATCH_1}${CMAKE_MATCH_2})")
-endforeach()
-if (CPACK_DEBIAN_PACKAGE_SHLIBS_LIST)
-  string(REPLACE ";" "\n" CPACK_DEBIAN_PACKAGE_SHLIBS "${CPACK_DEBIAN_PACKAGE_SHLIBS_LIST}")
+if(CPACK_DEBIAN_PACKAGE_GENERATE_SHLIBS AND READELF_PROG)
+  foreach(_FILE ${CPACK_DEB_SHARED_OBJECT_FILES})
+    extract_soname(${_FILE} libname soversion)
+    if(libname AND soversion)
+      list(APPEND CPACK_DEBIAN_PACKAGE_SHLIBS_LIST
+           "${libname} ${soversion} ${CPACK_DEBIAN_PACKAGE_NAME} (${CPACK_DEBIAN_PACKAGE_GENERATE_SHLIBS_POLICY} ${CPACK_PACKAGE_VERSION})")
```

Comment:
Wouldn't `CPACK_PACKAGE_VERSION` contain all of the version suffixes like build # etc.? I guess you'd better extract the version from the library extension.